### PR TITLE
New model branch

### PR DESCRIPTION
**Pull request by jalwin**

### CODE REVIEW

File: PokemonApp/CoreDataLayer/CoreDataPokemonProject.xcdatamodeld/CoreDataPokemon.xcdatamodel/contents [hunk 1]
```diff
@@ -3,15 +3,9 @@
     <entity name="Person" representedClassName="Person" syncable="YES">
         <attribute name="gender" optional="YES" attributeType="String" syncable="YES"/>
         <attribute name="name" optional="YES" attributeType="String" syncable="YES"/>
-        <relationship name="pokemon" optional="YES" toMany="YES" deletionRule="Nullify" destinationEntity="Pokemon" inverseName="person" inverseEntity="Pokemon" syncable="YES"/>
```
Comment (prrenfrew) [inline PokemonApp/CoreDataLayer/CoreDataPokemonProject.xcdatamodeld/CoreDataPokemon.xcdatamodel/contents]: Why is this data model still here? Are we using the CoreDataPokemonProject data model anywhere?

File: PokemonApp/Model/ContextJSONDecoder.swift [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+            return
+        }
+        self.userInfo[key] = CoreDataManager.shareManager.getContext()
+    }
+}
+
+public extension CodingUserInfoKey {
+    // Helper property to retrieve the context
+    static let ContextKey = CodingUserInfoKey(rawValue: "managedObjectContext")
+}
```
Comment (prrenfrew) [inline PokemonApp/Model/ContextJSONDecoder.swift]: Put this extension into another file.

File: PokemonApp/Model/Pokedex/Pokedex+CoreDataClass.swift [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        guard let codingUserInfoKeyManagedObjectContext = CodingUserInfoKey.ContextKey,
+            let managedObjectContext = decoder.userInfo[codingUserInfoKeyManagedObjectContext] as? NSManagedObjectContext,
+            let entity = NSEntityDescription.entity(forEntityName: "Pokedex", in: managedObjectContext) else {
+                fatalError("Failed to decode Pokedex")
+        }
+        
+        self.init(entity: entity, insertInto: managedObjectContext)
+        
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        self.caught = try container.decodeIfPresent(Int16.self, forKey: .caught) ?? 1
```
Comment (prrenfrew) [inline PokemonApp/Model/Pokedex/Pokedex+CoreDataClass.swift]: What is the purpose of `seen` and `caught`? At a minimum, the name of the properties needs to be more descriptive.

File: PokemonApp/Model/Pokemon/Pokemon+CoreDataClass.swift [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        guard let codingUserInfoKeyManagedObjectContext = CodingUserInfoKey.ContextKey,
+            let managedObjectContext = decoder.userInfo[codingUserInfoKeyManagedObjectContext] as? NSManagedObjectContext,
+            let entity = NSEntityDescription.entity(forEntityName: "Pokemon", in: managedObjectContext) else {
+                fatalError("Failed to decode Pokemon")
+        }
+        
+        self.init(entity: entity, insertInto: managedObjectContext)
+        
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        self.gender = try container.decodeIfPresent(Bool.self, forKey: .gender) ?? true
```
Comment (prrenfrew) [inline PokemonApp/Model/Pokemon/Pokemon+CoreDataClass.swift]: How does gender relate to a boolean?

File: PokemonApp/PokemonApp/AppDelegate.swift [hunk 5]
```diff
@@ -7,36 +7,83 @@
...
 
-  func applicationWillTerminate(_ application: UIApplication) {
+    func applicationWillTerminate(_ application: UIApplication) {
     // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
-  }
+    }
+    
+    // MARK: - Core Data stack
+    
+    lazy var persistentContainer: NSPersistentContainer = {
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/AppDelegate.swift]: Why was this added? CoreData layer is implemented elsewhere.

**Comment by prrenfrew** [comment]
Some of the initializers and functions added to managed objects should probably have some unit tests to make sure they are all working as expected.

File: PokemonApp/Model/Pokemon/Pokemon+CoreDataClass.swift (hunk 4 above)
Comment (jalwin) [reply PokemonApp/Model/Pokemon/Pokemon+CoreDataClass.swift]: male for true female for false

Comment (prrenfrew) [reply PokemonApp/Model/Pokemon/Pokemon+CoreDataClass.swift]: Seems like a good place for an enum.

File: PokemonApp/Model/ContextJSONDecoder.swift (hunk 2 above)
Comment (jalwin) [reply PokemonApp/Model/ContextJSONDecoder.swift]: DONE

File: PokemonApp/Model/Pokedex/Pokedex+CoreDataClass.swift (hunk 3 above)
Comment (jalwin) [reply PokemonApp/Model/Pokedex/Pokedex+CoreDataClass.swift]: On the pokedex, we want to display how many Pokemon were caught/seen

File: PokemonApp/CoreDataLayer/CoreDataPokemonProject.xcdatamodeld/CoreDataPokemon.xcdatamodel/contents (hunk 1 above)
Comment (jalwin) [reply PokemonApp/CoreDataLayer/CoreDataPokemonProject.xcdatamodeld/CoreDataPokemon.xcdatamodel/contents]: deleted file from project. should have been deleted but don't think Xcode deleted properly.

File: PokemonApp/Model/Pokemon/Pokemon+CoreDataClass.swift (hunk 4 above)
Comment (jalwin) [reply PokemonApp/Model/Pokemon/Pokemon+CoreDataClass.swift]: done

File: PokemonApp/PokemonApp/AppDelegate.swift (hunk 5 above)
Comment (jalwin) [reply PokemonApp/PokemonApp/AppDelegate.swift]: deleted and replaced with yours

File: PokemonApp/Model/ContextJSONDecoder.swift [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+import Foundation
+
+final class ContextJSONDecoder: JSONDecoder {
+    
+    override init() {
+        super.init()
+        guard let key = CodingUserInfoKey.ContextKey else {
+            return
+        }
```
Comment (prrenfrew) [inline PokemonApp/Model/ContextJSONDecoder.swift]: Make this guard one line.

File: PokemonApp/Model/ContextKey.swift [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+//
+//  Created by Alwin Amoros on 6/24/19.
+//  Copyright © 2019 PaulRenfrew. All rights reserved.
+//
+
+import Foundation
+
+public extension CodingUserInfoKey {
+    // Helper property to retrieve the context
+    static let ContextKey = CodingUserInfoKey(rawValue: "managedObjectContext")
```
Comment (prrenfrew) [inline PokemonApp/Model/ContextKey.swift]: Properties should use lowerCamelCase.

File: PokemonApp/Model/Pokedex/Pokedex+CoreDataClass.swift [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        case pokemonSeenInTotal
+        case pokemon
+    }
+    
+    // MARK: - Decodable
+    required convenience public init(from decoder: Decoder) throws {
+        guard let codingUserInfoKeyManagedObjectContext = CodingUserInfoKey.ContextKey,
+            let managedObjectContext = decoder.userInfo[codingUserInfoKeyManagedObjectContext] as? NSManagedObjectContext,
+            let entity = NSEntityDescription.entity(forEntityName: "Pokedex", in: managedObjectContext) else {
+                fatalError("Failed to decode Pokedex")
```
Comment (prrenfrew) [inline PokemonApp/Model/Pokedex/Pokedex+CoreDataClass.swift]: Any reason you did fatalError over throwing an error?

File: PokemonApp/Model/Pokemon/Pokemon+CoreDataClass.swift [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        case pokemonCaughtTotal
+        case pokemonDescription
+    }
+    
+    // MARK: - Decodable
+    required convenience public init(from decoder: Decoder) throws {
+        guard let codingUserInfoKeyManagedObjectContext = CodingUserInfoKey.ContextKey,
+            let managedObjectContext = decoder.userInfo[codingUserInfoKeyManagedObjectContext] as? NSManagedObjectContext,
+            let entity = NSEntityDescription.entity(forEntityName: "Pokemon", in: managedObjectContext) else {
+                fatalError("Failed to decode Pokemon")
```
Comment (prrenfrew) [inline PokemonApp/Model/Pokemon/Pokemon+CoreDataClass.swift]: FatalError over throwing error?

File: PokemonApp/Model/Pokemon/PokemonStats/PokemonStat+CoreDataClass.swift [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+        case pokemonStatURL = "url"
+        case pokemonStats
+    }
+    
+    // MARK: - Decodable
+    required convenience public init(from decoder: Decoder) throws {
+        guard let codingUserInfoKeyManagedObjectContext = CodingUserInfoKey.ContextKey,
+            let managedObjectContext = decoder.userInfo[codingUserInfoKeyManagedObjectContext] as? NSManagedObjectContext,
+            let entity = NSEntityDescription.entity(forEntityName: "PokemonStat", in: managedObjectContext) else {
+                fatalError("Failed to decode PokemonStat")
```
Comment (prrenfrew) [inline PokemonApp/Model/Pokemon/PokemonStats/PokemonStat+CoreDataClass.swift]: FatalError over throwing error?

File: PokemonApp/Model/Pokemon/PokemonStats/PokemonStats+CoreDataClass.swift [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+        case pokemon
+        case pokemonStat = "stat"
+    }
+    
+    // MARK: - Decodable
+    required convenience public init(from decoder: Decoder) throws {
+        guard let codingUserInfoKeyManagedObjectContext = CodingUserInfoKey.ContextKey,
+            let managedObjectContext = decoder.userInfo[codingUserInfoKeyManagedObjectContext] as? NSManagedObjectContext,
+            let entity = NSEntityDescription.entity(forEntityName: "PokemonStats", in: managedObjectContext) else {
+                fatalError("Failed to decode PokemonStats")
```
Comment (prrenfrew) [inline PokemonApp/Model/Pokemon/PokemonStats/PokemonStats+CoreDataClass.swift]: FatalError over throwing error?

File: PokemonApp/Model/Pokemon/PokemonTypes/PokemonType+CoreDataClass.swift [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+        case pokemonTypes
+        case pokemonTypeURL = "url"
+    }
+    
+    // MARK: - Decodable
+    required convenience public init(from decoder: Decoder) throws {
+        guard let codingUserInfoKeyManagedObjectContext = CodingUserInfoKey.ContextKey,
+            let managedObjectContext = decoder.userInfo[codingUserInfoKeyManagedObjectContext] as? NSManagedObjectContext,
+            let entity = NSEntityDescription.entity(forEntityName: "PokemonType", in: managedObjectContext) else {
+                fatalError("Failed to decode PokemonType")
```
Comment (prrenfrew) [inline PokemonApp/Model/Pokemon/PokemonTypes/PokemonType+CoreDataClass.swift]: FatalError over throwing error?

File: PokemonApp/Model/Pokemon/PokemonTypes/PokemonTypes+CoreDataClass.swift [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+        case pokemon
+        case pokemonType = "type"
+    }
+    
+    // MARK: - Decodable
+    required convenience public init(from decoder: Decoder) throws {
+        guard let codingUserInfoKeyManagedObjectContext = CodingUserInfoKey.ContextKey,
+            let managedObjectContext = decoder.userInfo[codingUserInfoKeyManagedObjectContext] as? NSManagedObjectContext,
+            let entity = NSEntityDescription.entity(forEntityName: "PokemonTypes", in: managedObjectContext) else {
+                fatalError("Failed to decode PokemonTypes")
```
Comment (prrenfrew) [inline PokemonApp/Model/Pokemon/PokemonTypes/PokemonTypes+CoreDataClass.swift]: FatalError over throwing error?

File: PokemonApp/Model/Pokemon/Sprite/Sprite+CoreDataClass.swift [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+        case frontShinyPokemonURL = "front_shiny"
+        case pokemon
+    }
+
+    // MARK: - Decodable
+    required convenience public init(from decoder: Decoder) throws {
+        guard let codingUserInfoKeyManagedObjectContext = CodingUserInfoKey.ContextKey,
+            let managedObjectContext = decoder.userInfo[codingUserInfoKeyManagedObjectContext] as? NSManagedObjectContext,
+            let entity = NSEntityDescription.entity(forEntityName: "Sprite", in: managedObjectContext) else {
+                fatalError("Failed to decode Sprite")
```
Comment (prrenfrew) [inline PokemonApp/Model/Pokemon/Sprite/Sprite+CoreDataClass.swift]: FatalError over throwing error?

File: PokemonApp/Network Layer/NetworkController.swift [hunk 15]
```diff
@@ -4,6 +4,8 @@
 //
 //  Created by MAC Consultant on 6/10/19.
 //  Copyright © 2019 PaulRenfrew. All rights reserved.
+//  Revised: Alwin 6/21/2019
+//  Modified the JSONDecoder, had to subclass it so I can add the ContextKey and get the ContextObjectManager
```
Comment (prrenfrew) [inline PokemonApp/Network Layer/NetworkController.swift]: The beginning comments are for file documentation. This should have bee the commit message when you committed the mentioned change.

File: PokemonApp/PokemonApp/Info.plist [hunk 16]
```diff
@@ -4,6 +4,10 @@
 <dict>
 	<key>CFBundleDevelopmentRegion</key>
 	<string>$(DEVELOPMENT_LANGUAGE)</string>
+	<key>CFBundleDocumentTypes</key>
+	<array>
+		<dict/>
+	</array>
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/Info.plist]: What is this change for?

File: PokemonApp/PokemonAppTests/PokemonModelTest.swift [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+
+    
+    override func setUp() {
+        super.setUp()
+        sut = NetworkController()
+        readJson()
+        context = CoreDataManager.shareManager.getContext()
+    }
+    
+    private func readJson() {
```
Comment (prrenfrew) [inline PokemonApp/PokemonAppTests/PokemonModelTest.swift]: What is the point of this function? Doesn't `pokemonData` cover everything that this does?

File: PokemonApp/PokemonAppTests/PokemonModelTest.swift [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+        super.tearDown()
+    }
+    
+    func testDecodingPokemonModel() {
+        let mockSession = URLMockSession(data: pokemonData, error: nil)
+
+        sut.getData(session: mockSession, url: validURL) { (ditto: Pokemon?, error) in
+            XCTAssertNotNil(ditto)
+            XCTAssertEqual(ditto?.pokemonName, "ditto")
+            //print(ditto?.pokemonTypes.first?.pokemonType?.pokemonTypeName)
```
Comment (prrenfrew) [inline PokemonApp/PokemonAppTests/PokemonModelTest.swift]: Remove commented code.

File: PokemonApp/PokemonAppTests/PokemonModelTest.swift [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+            self.pokemons = try! CoreDataManager.shareManager.fetchAll(entity: Pokemon.self, storeType: NSInMemoryStoreType)
+            
+            for pokemon in self.pokemons {
+                try! CoreDataManager.shareManager.cleanRemove(entity: pokemon)
+            }
+            try! CoreDataManager.shareManager.saveContext()
+            
+            //  verifying that the item is removed.
+            print(self.pokemons.count)
+            print(self.pokemons.first?.pokemonName)
```
Comment (prrenfrew) [inline PokemonApp/PokemonAppTests/PokemonModelTest.swift]: Remove debug print statements. If you want to verify something, use XCTAssert

File: PokemonApp/CoreDataLayer/CoreDataManager.swift [hunk 20]
```diff
@@ -22,7 +23,8 @@ final class CoreDataManager {
         try context.save()
     }
     
-    private func getContext(storeType: String = NSSQLiteStoreType) -> NSManagedObjectContext {
+    // Changed this to regular since getting the ViewContext needed for Model and Core data was Inaccessible
```
Comment (prrenfrew) [inline PokemonApp/CoreDataLayer/CoreDataManager.swift]: `regular` is inaccurate. You changed it to internal, which is the default, so you could access it.

File: PokemonApp/Model/Pokedex/Pokedex+CoreDataClass.swift [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+                throw ModelError.couldNotDecodePokedex
+                
+        }
+        
+        self.init(entity: entity, insertInto: managedObjectContext)
+        
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        self.pokemonCaughtInTotal = try container.decodeIfPresent(Int16.self, forKey: .pokemonCaughtInTotal) ?? 1
+        self.pokemonSeenInTotal = try container.decodeIfPresent(Int16.self, forKey: .pokemonSeenInTotal) ?? 1
+        let pokemonArray = try container.decode([Pokemon].self, forKey: .pokemon)
```
Comment (prrenfrew) [inline PokemonApp/Model/Pokedex/Pokedex+CoreDataClass.swift]: Have you tried to do `Set<Pokemon>.self` as opposed to `[Pokemon].self`? This would allow us to set `pokemon` directly.

File: PokemonApp/Model/Pokemon/PokemonTypes/PokemonType+CoreDataClass.swift (hunk 12 above)
Comment (prrenfrew) [reply PokemonApp/Model/Pokemon/PokemonTypes/PokemonType+CoreDataClass.swift]: Still crashing app as opposed to throwing error.

File: PokemonApp/Model/Pokemon/PokemonTypes/PokemonTypes+CoreDataClass.swift (hunk 13 above)
Comment (prrenfrew) [reply PokemonApp/Model/Pokemon/PokemonTypes/PokemonTypes+CoreDataClass.swift]: Still crashing app over throwing error.

File: PokemonApp/PokemonApp/AppDelegate.swift [hunk 22]
```diff
@@ -10,33 +10,37 @@ import UIKit
...
-  }
-
-  func applicationWillTerminate(_ application: UIApplication) {
-    // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
-  }
+    
+    var window: UIWindow?
+    
+    
+    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/AppDelegate.swift]: Why did you make these changes? As per the style guide, indentation should only be 2 spaces.

File: PokemonApp/PokemonAppTests/PokemonModelTest.swift [hunk 23]
```diff
@@ -0,0 +1,50 @@
+import XCTest
+import CoreData
+@testable import PokemonApp
+
+final class PokemonModelTest: XCTestCase {
+    
+    let validURL = "https://pokeapi.co/api/v2/pokemon/ditto/"
+    var sut: NetworkController!
```
Comment (prrenfrew) [inline PokemonApp/PokemonAppTests/PokemonModelTest.swift]: `sut` means system under test. Are you testing the NetworkController?

File: PokemonApp/PokemonAppTests/PokemonModelTest.swift [hunk 24]
```diff
@@ -0,0 +1,50 @@
...
+        return data
+    }
+    var pokemons: [Pokemon]!
+    var context: NSManagedObjectContext!
+
+    
+    override func setUp() {
+        super.setUp()
+        sut = NetworkController()
+        context = CoreDataManager.shareManager.getContext()
```
Comment (prrenfrew) [inline PokemonApp/PokemonAppTests/PokemonModelTest.swift]: Is there a reason why you aren't using the in memory type for the persistent store here?

File: PokemonApp/Model/Pokemon/Pokemon+CoreDataClass.swift [hunk 25]
```diff
@@ -0,0 +1,64 @@
...
+        }
+        
+        self.init(entity: entity, insertInto: managedObjectContext)
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        self.gender = try container.decodeIfPresent(Int16.self, forKey: .gender) ?? PokemonGender.Male.rawValue
+        self.pokemonName = try container.decode(String.self, forKey: .pokemonName)
+        self.pokemonSeenTotal = try container.decodeIfPresent(Int16.self, forKey: .pokemonSeenTotal) ?? 0
+        self.pokemonNumber = try container.decode(Int16.self, forKey: .pokemonNumber)
+        self.pokemonCaughtTotal = try container.decodeIfPresent(Int16.self, forKey: .pokemonCaughtTotal) ?? 0
+        let statsArray = try container.decode([PokemonStats].self, forKey: .pokemonStats)
```
Comment (prrenfrew) [inline PokemonApp/Model/Pokemon/Pokemon+CoreDataClass.swift]: Set over Array?

File: PokemonApp/PokemonApp/Base.lproj/Main.storyboard [hunk 26]
```diff
@@ -1,6 +1,10 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <document type="com.apple.InterfaceBuilder3.CocoaTouch.Storyboard.XIB" version="3.0" toolsVersion="14490.70" targetRuntime="iOS.CocoaTouch" propertyAccessControl="none" useAutolayout="YES" useTraitCollections="YES" useSafeAreas="YES" colorMatched="YES" initialViewController="PF2-HB-Htm">
+    <device id="retina6_1" orientation="portrait">
+        <adaptation id="fullscreen"/>
+    </device>
```
Comment (prrenfrew) [inline PokemonApp/PokemonApp/Base.lproj/Main.storyboard]: What was the purpose for these changes?
